listenOnce fixes

jamespthomas opened the pull request:
Hi!

There were some situations where listenOnce calls would crash (for example, when doing something like irc.names() inside a connect function), by adding a unique id (single_listener_count) to the internal storage hashmap it now works.

Also added a check to make sure it didn't attempt to remove a listener it had already removed, this would happen if there were multiple listenOnce's happening on the same event (In my case, calling .names inside a for loop on a channel array)

Hope you find them useful!

Thanks
James

jamespthomas committed:
fixed: Added event identifier to listenOnce hashmap entries to prevent crash

jamespthomas committed:
fixed: check to make sure we haven't already deleted the listener!

gf3 commented:
That's excellent! Code looks great as well, thank you. I have one small request, just so your additions match the coding styles of the rest of the package, could you change things like `[hollaback+"_"+listenId]` to `[ hollaback + '_' + listenId ]`? Thanks again!

jamespthomas committed:
Code cleanup, also fixed so it actually works!

jamespthomas commented:
Hi, fixed the code now, only using listenId for the key now (since that's unique enough!), also fixed a massive bug in add listener, don't know how that got in there! Testing and working fine now.

jamespthomas committed:
fixed: this.names was not returning the channel name

jamespthomas committed:
fixed: listenOnce listeners incorrectly calling callbacks when there was more than one event of the same kind

jamespthomas commented:
Hello,

Another problem I had (with names again) was if you added several listenOnce listeners with the same event, before any response came back/was processed (so a 353) it would call each hollaback with the same message. I added some more parameters to listenOnce to prevent this. 

If you call it with "listenFor" (some string to be matched) and "listenParam" (the message param to match in), it will now check to see if these requirements are met before calling the holla (of course, if these aren't sent the callback for the event is called). Only if these conditions are met is the listener removed.

So the listenOnce call for names is now listenOnce( '353', function(), channel, 2 ) which will match the channel name on message.params[2] before calling the holla and removing the listener.

Thanks
James

gf3 commented:
I am going to switch to the internal listenOnce.
